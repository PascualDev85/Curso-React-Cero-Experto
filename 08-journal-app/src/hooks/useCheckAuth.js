import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { onAuthStateChanged } from "firebase/auth";

import { FirebaseAuth } from "../firebase/config";
import { login, logout } from "../store/auth";
import { startLoadingNotes } from "../store/journal";

export const useCheckAuth = () => {
  // mostrar un spinner cuando se está checking el status del usuario
  const { status } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  // estar pendiente de la autenticación
  useEffect(() => {
    // FirebaseAuth observardor (función que está emitiendo valores)
    onAuthStateChanged(FirebaseAuth, async (user) => {
      // console.log(user);

      // si no hay ningún usuario llamo al logout
      if (!user) return dispatch(logout());

      // si hay un usuario llamo al login
      const { uid, email, displayName, photoURL } = user;
      dispatch(login({ uid, email, displayName, photoURL }));
      // cargar una nota
      dispatch(startLoadingNotes());
    });
  }, []);

  //   return {
  //     status,
  //   };

  return status;
};
