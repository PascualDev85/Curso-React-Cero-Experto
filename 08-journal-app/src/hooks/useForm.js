import { useEffect, useState } from "react";

export const useForm = (initialForm = {}, formValidations = {}) => {
  const [formState, setFormState] = useState(initialForm);
  const [formValidation, setFormValidation] = useState({});

  // cada vez que cambie el formstate se vuelve a activar el createValidator
  useEffect(() => {
    createValidators();
  }, [formState]);

  const onInputChange = ({ target }) => {
    const { name, value } = target;
    setFormState({
      ...formState,
      [name]: value,
    });
  };

  const onResetForm = () => {
    setFormState(initialForm);
  };

  const createValidators = () => {
    const formCheckedValues = {};

    // iterar cada una de las validaciones que tengo en Register
    for (const formField of Object.keys(formValidations)) {
      // console.log(formField);
      // destructuring función de validaciones y errorMessage
      const [fn, errorMessage] = formValidations[formField];

      // esto quiero decir que comprobamos todas las validaciones que estan fn si es correcto será null y si no será el mensaje de error
      formCheckedValues[`${formField}Valid`] = fn(formState[formField])
        ? null
        : errorMessage;
    }

    setFormValidation(formCheckedValues);
  };

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm,

    ...formValidation,
  };
};
