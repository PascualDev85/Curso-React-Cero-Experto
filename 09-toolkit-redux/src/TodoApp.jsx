import { useState } from "react";
import { useGetTodoByIdQuery, useGetTodosQuery } from "./store/apis";

export const TodoApp = () => {
  //   const { data: todos = [], isLoading } = useGetTodosQuery();
  //   console.log(algo);

  const [todoId, setTodoId] = useState(1);
  const { data: todo, isLoading } = useGetTodoByIdQuery(todoId);
  //   console.log(todo);

  const nextTodo = () => {
    setTodoId(todoId + 1);
  };

  const prevTodo = () => {
    if (todoId === 1) return false;
    setTodoId(todoId - 1);
  };

  return (
    <>
      <h1>Todos - RFT Query</h1>
      <hr />
      <h4>isLoading: {isLoading ? "True" : "False"}</h4>

      <pre> {JSON.stringify(todo)} </pre>

      {/* <ul>
        {todos.map((todo) => (
          <li key={todo.id}>
            <strong>{todo.completed ? "DONE " : "Pending "}</strong>
            {todo.title}
          </li>
        ))}
      </ul> */}
      <button onClick={prevTodo}>Prev Todo</button>
      <button onClick={nextTodo}>Next Todo</button>
    </>
  );
};
