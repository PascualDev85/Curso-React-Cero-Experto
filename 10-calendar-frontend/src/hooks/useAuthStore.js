import { useDispatch, useSelector } from "react-redux";
import calendarApi from "../api/calendarApi";
import {
  clearErrorMessage,
  onChecking,
  onLogin,
  onLogout,
  onLogoutCalendar,
} from "../store";

export const useAuthStore = () => {
  const { status, user, errorMessage } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {
    // console.log({ email, password });
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth", { email, password });
      //   console.log({ response });
      // establecemos el token en el localStorage
      localStorage.setItem("token", data.token);
      // establecemos el token en la fecha actual. Para ahorrar una petición al backend mientras el token sea válido
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      //   console.log(error);
      dispatch(onLogout("Email y/o password son incorrectas"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const startRegister = async ({ email, password, name }) => {
    dispatch(onChecking());
    try {
      const { data } = await calendarApi.post("/auth/new", {
        email,
        password,
        name,
      });

      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      //   console.log(error);
      dispatch(onLogout(error.response.data?.msg || "Error en el registro"));
      setTimeout(() => {
        dispatch(clearErrorMessage());
      }, 10);
    }
  };

  const checkAuthToken = async () => {
    const token = localStorage.getItem("token");
    if (!token) return dispatch(onLogout());

    try {
      const { data } = await calendarApi.get("/auth/renew");
      //   console.log({ data });
      // TODO: crear una función para refactorear este código
      localStorage.setItem("token", data.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(onLogin({ name: data.name, uid: data.uid }));
    } catch (error) {
      //   console.log(error);
      localStorage.clear();
      dispatch(onLogout("El token ha expirado"));
    }
  };

  const startLogout = () => {
    localStorage.clear();
    dispatch(onLogoutCalendar());
    dispatch(onLogout());
  };

  return {
    //* Propiedades
    status,
    user,
    errorMessage,

    //* Métodos
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
