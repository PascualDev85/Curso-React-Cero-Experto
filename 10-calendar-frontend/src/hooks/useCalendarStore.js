import { useDispatch, useSelector } from "react-redux";
import Swal from "sweetalert2";
import calendarApi from "../api/calendarApi";
import { convertEventsToDateEvents } from "../helpers";
import {
  onAddNewEvent,
  onDeleteEvent,
  onLoadEvents,
  onSetActiveEvent,
  onUpdateEvent,
} from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {
  const dispatch = useDispatch();
  // slice de calendar
  const { events, activeEvent } = useSelector((state) => state.calendar);
  // slice de auth
  const { user } = useSelector((state) => state.auth);

  // activar evento
  const setActiveEvent = (calendarEvent) => {
    dispatch(onSetActiveEvent(calendarEvent));
  };

  // guardar evento
  const startSavingEvent = async (calendarEvent) => {
    //Todo sale bien

    try {
      if (calendarEvent.id) {
        // actualizar evento ennviar el evento al backend
        await calendarApi.put(`/events/${calendarEvent.id}`, calendarEvent);
        // actualizar evento haciendo un dispatch
        dispatch(onUpdateEvent({ ...calendarEvent, user }));
        return;
      }
      //Creando un nuevo evento
      // aqui se hace la petición al backend que está en el body
      await calendarApi.post("/events", calendarEvent);
      // console.log({ data });

      dispatch(
        onAddNewEvent({
          //Creando un nuevo evento
          ...calendarEvent,
          // _id: new Date().getTime(), // se elimina porque ya se genera en el backend
          id: data.evento.id,
          user,
        })
      );
      // Si hay un error
    } catch (error) {
      console.log("Error al guardar el evento", error);
      Swal.fire("Error al guardar", error.response.data.msg, "error");
    }
  };

  // eliminar evento
  const startDeletingEvent = async () => {
    // Todo: Llegar al backend
    try {
      // llamar al backend para eliminar el evento
      await calendarApi.delete(`/events/${activeEvent.id}`);
      // llamar al dispatch para eliminar el evento
      dispatch(onDeleteEvent());
    } catch (error) {
      console.log("Error al eliminar el evento", error);
      Swal.fire("Error al eliminar", error.response.data.msg, "error");
    }
  };

  // cargar eventos del backend
  const startLoadingEvents = async () => {
    try {
      // llamar al backend para obtener los eventos
      const { data } = await calendarApi.get("/events");
      // console.log({ data });
      // convertir la fecha de los eventos
      const events = convertEventsToDateEvents(data.eventos);
      // console.log({ events });
      // cargar los eventos  que se obtienen del backend
      dispatch(onLoadEvents(events));
    } catch (error) {
      console.log("Error al cargar los eventos", error);
    }
  };

  return {
    //* Propiedades
    events,
    activeEvent,
    // nueva propiedad para saber si hay un evento activo
    hasEventSelected: !!activeEvent,
    // doble negacion para convertir a boolean, si no hay evento, devuelve false y si hay, devuelve true. De esta ocultaremos el boton de eliminar evento

    //* Metodos
    setActiveEvent,
    startSavingEvent,
    startDeletingEvent,
    startLoadingEvents,
  };
};
