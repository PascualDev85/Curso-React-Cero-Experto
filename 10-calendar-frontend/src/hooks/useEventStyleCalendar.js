// hacer un custom hook para obtener el estilo de los eventos

export const useEventStyleCalendar = (event, start, end, isSelected) => {
  //   console.log(event, start, end, isSelected);

  const style = {
    backgroundColor: "#367CF7",
    borderRadius: "0px",
    opacity: 0.9,
    display: "block",
    color: "white",
  };

  return {
    style,
  };
};
