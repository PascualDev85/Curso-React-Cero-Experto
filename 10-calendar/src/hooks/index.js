export * from "./useEventStyleCalendar";
export * from "./useUiStore";
export * from "./useCalendarStore";
