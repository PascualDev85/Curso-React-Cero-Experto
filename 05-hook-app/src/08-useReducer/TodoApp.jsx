import { useTodos } from "../hooks/";
import { TodoAdd, TodoList } from "./components/";

export const TodoApp = () => {
  const {
    todos,
    handleNewTodo,
    handleDeleteTodo,
    handleToggleTodo,
    todosCount,
    pendingTodosCount,
  } = useTodos();

  return (
    <>
      <h1>
        TodoApp {todosCount}, <small>pendientes: {pendingTodosCount}</small>
      </h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            onDeleteTodo={handleDeleteTodo}
            onToggleTodo={handleToggleTodo}
          />
        </div>
        <div className="col-5">
          <h4>Agregar TODO</h4>
          <hr />
          <TodoAdd
            // recibir recibiría el TODO, función de flecha le pasaría el TODO
            // onNewTodo={(todo) => handleNewTodo(todo)}
            // o también se puede pasar directamente la función en el mismo orden este caso
            onNewTodo={handleNewTodo}
          />
        </div>
      </div>
    </>
  );
};
