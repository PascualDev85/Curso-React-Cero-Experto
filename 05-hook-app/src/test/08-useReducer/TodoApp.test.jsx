import { render, screen } from "@testing-library/react";
import { TodoApp } from "../../08-useReducer/TodoApp";
import { useTodos } from "../../hooks/useTodos";

jest.mock("../../hooks/useTodos");

describe("Pruebas en <TodoAdd />", () => {
  useTodos.mockReturnValue({
    todos: [
      {
        id: 1,
        description: "Aprender React",
        done: false,
      },
      {
        id: 2,
        description: "Aprender Testing",
        done: true,
      },
    ],
    handleNewTodo: jest.fn(),
    handleDeleteTodo: jest.fn(),
    handleToggleTodo: jest.fn(),
    todosCount: 2,
    pendingTodosCount: 1,
  });

  test("debe hacer el match con el snapshot", () => {
    const { container } = render(<TodoApp />);
    expect(container).toMatchSnapshot();
  });

  test("debe de mostrar el componente correctamente ", () => {
    render(<TodoApp />);

    expect(screen.getByText("Aprender React")).toBeTruthy();
    expect(screen.getByText("Aprender Testing")).toBeTruthy();
    expect(screen.getByRole("textbox")).toBeTruthy();

    // console.log(screen.getByRole("textbox").name);

    // screen.debug();
  });
});
