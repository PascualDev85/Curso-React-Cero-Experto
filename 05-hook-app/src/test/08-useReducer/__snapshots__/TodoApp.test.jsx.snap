// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> debe hacer el match con el snapshot 1`] = `
<div>
  <h1>
    TodoApp 
    2
    , 
    <small>
      pendientes: 
      1
    </small>
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center "
          >
            Aprender React
          </span>
          <button
            aria-label="deleteButton"
            class="btn btn-danger mx-3"
          >
            Borrar
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center text-decoration-line-through"
          >
            Aprender Testing
          </span>
          <button
            aria-label="deleteButton"
            class="btn btn-danger mx-3"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar TODO
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="description"
          placeholder="Introdcir tarea"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2 btn-block"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
